src/textual/widget.py

```diff
@@ -333,6 +335,44 @@ def offset(self) -> Offset:
     def offset(self, offset: Offset) -> None:
         self.styles.offset = ScalarOffset.from_offset(offset)
 
+    def get_child_by_id(self, id: str) -> Widget:
+        """Return the first child (immediate descendent) of this node with the given ID.
+
+        Args:
+            id (str): The ID of the child.
+
+        Returns:
+            DOMNode: The first child of this node with the ID.
+
+        Raises:
+            NoMatches: if no children could be found for this ID
+        """
+        child = self.children._get_by_id(id)
+        if child is not None:
+            return child
+        raise NoMatches(f"No child found with id={id!r}")
+
+    def get_widget_by_id(self, id: str) -> Widget:
+        """Return the first descendant widget with the given ID.
+        Performs a breadth-first search rooted at this node, but this node won't match
+        if it has a matching id.
+
+        Args:
+            id (str): The ID to search for in the subtree
+
+        Returns:
+            DOMNode: The first descendant encountered with this ID.
+
+        Raises:
+            NoMatches: if no children could be found for this ID
+        """
+        for child in self.walk_children(method="depth"):
```

Comment:
This will walk the entire DOM since `walk_children` returns a list.

so performance will unfortunately be much the same. I think we might need a more specialised way of walking children for this method.

I'm also think that the walk should be breadth first. If there happens to be multiple IDs, I think you would expect the most immediate matching node to be returned.

Reply:
Scratch that. I'm thinking you were right the first time in `"depth"`. Other wise `query_one` and `get_widget_by_id` could return different values.

Reply:
Have updated to use `walk_depth_first`